### ISSUE TITLE
Parameters for modules

### ISSUE THREAD

**Issue by acfoltzer** [GaloisInc/cryptol#365]
**edited with revised syntax proposal**
## The `include` pattern

We'd like a way for modules to accept type and value
parameters. Currently, we have the `include` pattern:

`Foo.cry`:

```
f : [n] -> [n]
f x = x + 1
```

`Bar.cry`:

```
module Bar where

type n = 8

include "Foo.cry"

answer = f 120
// answer = 121
```

Let's call `Foo` what it is: a module fragment. It cannot be loaded
directly, since it has the free type variable `n`, and since it's not
`Main.cry` its module name will not match its filename. Instead it can
be included into a module such as `Bar` which binds its free variables
in order to comprise a complete module.

This is awkward for a number of reasons:
- `Foo` cannot be loaded by itself
- Instantiating `Foo` requires that you make another file such as
  `Bar.cry`; you cannot instantiate it at the REPL
- Modules with an explicit `module` declaration cannot be included as
  module fragments due to the invalid syntax that would be introduced
- Names that are private in a module fragment are in scope in the
  modules that include them
- A single module cannot instantiate a multiple fragment more than
  once with different parameters

---
## Parameterized module proposal

Revising the example above with strawman syntax for parameterizing
modules, rather than relying on textual inclusion:

`Foo.cry`:

```
module Foo
  {n : #, m : #} (fin n, fin m, n > m)
where

f : (fin n) => [n] -> [n]
f x = x + 1

g : [m] -> [m]
g y = y
```

`Bar.cry`:

```
module Bar where

import Foo { n = 16, m = 8 }

answer = f 120
// answer = 121
```

We augment the syntax of the `module` declaration to can contain
`schema_vars` and `schema_quals` for the type parameters. We then
instantiate the module using record syntax similar to the explicit
type instantiation syntax

With this basic functionality in hand, we can imagine some useful
extensions:
### Instantiating multiple times

I think we would get this for free, or very cheap.

`Bar.cry`:

```
module Bar where

import Foo { n = 8,  m = 4 } as FooA
import Foo { n = 16, m = 8 } as FooB

answerA = FooA::f 120
// answerA : [8]
answerB = FooB::f 120
// answerB : [16]
```
### Default parameters

Rather than require the parameters to be instantiated every time, we
introduce a new declaration form for default parameters (again,
strawman syntax):

`Foo.cry`:

```
module Foo 
  {n} (fin n)
where

default type n = 32

f : [n] -> [n]
f x = x + 1
```

`Bar.cry`:

```
module Bar where

import Foo

answer = f 120
// answer : [32]
```

If we provide an argument when instantiating, it overrides the
default, otherwise the default is used.
#### Bonus

Could this same declaration be used on other polymorphic type
variables to help wrangle defaulting?
### Scoping of instantiation arguments

Even trickier is the case where an RHS of a default or a parameter
instantiation contains names defined in modules that are imported, or
that are themselves parameters. We definitely want to support this,
otherwise the prelude's type synonyms wouldn't work. To add even more
difficulty, what should happen if an RHS contains names defined in
_parametric_ modules that are imported?  What if it gets even worse:

```
module Foo
  {xs}
where

import Baz  { ys = xs } // defines q
import Quux { p  = q  }

default type xs = Word // Word defined in Prelude
```

There are a few levels of support we could pursue here. We should
figure out what the sweet spot is to support the common use
cases. (**edit**: this probably is a non-issue with the syntactic
elaboration approach)
### Loading uninstantiated modules

This might be on the critical path anyway if we want to use the
syntacic elaboration approach to implementing parameters.

`Foo.cry`:

```
module Foo
  {n} (fin n)
where

f : [n] -> [n]
f x = x + 1
```

`Bar.cry`:

```
module Bar where

import Foo // no instantiation

answer = f`{n=8} 120
// answer = 121
```

```
Cryptol> :l Foo.cry
Foo> :t f
f : {n} (fin n) => [n] -> [n]
Foo> f 120 : [8]
121
Foo> f 121 : [32]
121
```
### Value parameters

This might be a lot more complicated depending on how we implement the
basic parametric feature. A syntactic approach might make this
relatively straightforward. (**edit**: this uses the initial
`parameters` syntax because it's not clear how to nicely add on to the
`module` style declaration above)

`Foo.cry`:

```
module Foo where

parameters
  type {n} (fin n)
  c : [n]
  default c = 1

f : [n] -> [n]
f x = x + c
```

`Bar.cry`:

```
module Bar where

import Foo { type n = 8, c = 42 }

answer = f 120
// answer = 162
```

Essentially, we augment the `parameters` block with type signatures
and default bindings for value parameters, and then use the `type`
keyword when instantiating in order to distinguish between type and
value parameters.

**Comment by acfoltzer** [comment]
cc @elliottt

**Comment by brianhuffman** [comment]
@acfoltzer mentioned a "syntactic elaboration approach" to implementing parameterized modules; I'll explain that in some more detail here.

Say we have a parameterized module Foo.cry:
```
module Foo {n} (fin n) where
f : [n] -> [n]
f x = x + 1
type list a = [n]a
```

and a client module Bar.cry:
```
module Bar where
import Foo {n=bar}
type bar = 8
g : [bar]
g = f zero
h : {a} list a
h = zero
```

First, we will compile module `Foo` by converting it to an ordinary module, adding type parameters to every function and type synonym:
```
module Foo where
f : {n} (fin n) [n] -> [n]
f x = x + 1
type list n a = [n]a
```

Next, we compile the client module `Bar` by translating the parameterized import to an ordinary (qualified) import plus a bunch of new declarations:
```
module Bar where
import Foo as Foo_raw

/* Instantiated declarations from Foo */
f : [bar] -> [bar]
f = Foo_raw::f`{n=bar}
type list a = Foo_raw::list n a

/* unchanged declarations from Bar */
type bar = 8
g : [bar]
g = f zero
h : {a} list a
h = zero
```

Dependencies among the declarations in the body of the client module `Bar` and the instantiation parameters would then be handled by the typechecker in the normal way.

**Comment by yav** [comment]
We have discussed variations on this for a long time, perhaps it is time to revisit it again.  Here is some background to consider.  

To make the best design choices, we should consider what are the motivations for using the `#include` pattern.  I can think of two:

  1. factor out common parameters to de-clutter functions and local type signatures, 
  2. get a program to type check in some concrete cases, when the general would be rejected, usually due to weaknesses in the constraint solver.

I think that 1 is nice, and this proposal (and variants, that we can bike-shed about later), would help with it.

Unfortunately, I think that 2 is the much more common reason to use `#include` and having these parameters does not really help us there.   The issue is that sometimes Cryptol's type-checker is not smart enough to solve a problem in general, but it can solve it for concrete instances.  Which leaves us with a dilemma: do we check parametrized modules, or do we delay the checking until module instantiation.   Delaying checking until instantiation will do what `#include` does, but it is not great as it reports errors too late---a type error in a library will be discovered only when a user tries to use the library.  OTOH, trying to type-check an un-instantiated parametrized module will fail in the same way it fails currently, as the problem Cryptol would be solving is not any easier.

So, given that parametrized modules will not really help with type checking, we'd have to decide if the benefits of 1) above are greater than the cost of making the language and implementation more complex.  I've sometimes wanted this sort of thing but, for example, Haskell does not provide this feature either, and one can write a lot of code without this being too much of an issue.

**Comment by acfoltzer** [comment]
At least in this round of discussions, I don't think I heard of anyone using `include` for reason 2. The quality-of-life aspects yielded seem to be worth it alone given how folks are using `include` today.

**Comment by brianhuffman** [comment]
The parameterized module feature was merged into master on Oct 27, 2017 (07eac4798926d760d58f24f59df558b0f88c1657).
